### PR TITLE
[clang-tidy] bugprone-assert-side-effect non-const operator methods

### PR DESCRIPTION
**Pull request by schenker**
With this PR, `bugprone-assert-side-effect` reports non-const calls to `<<` and `>>` operators and does no longer consider any const operator calls to have side effects. 

E.g. the following snippet is now reported and was previously not:
```
std::stringstream ss;
assert(ss << 1);
```

The following snippet was previously a false-positive and is not reported anymore:
```
struct {
  int operator+=(int i) const { return i; }
} t;
assert(t += 1);
```

### CODE REVIEW

**Review by PiotrZSL — APPROVED** [review]
Update documentation of this check to mention this, and update release notes.
Code looks fine.

File: clang-tools-extra/clang-tidy/bugprone/AssertSideEffectCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/bugprone/AssertSideEffectCheck.cpp:46]: should it be `if (MethodDecl->isConst()) return false`; otherwise code in line 49 could be a dead code.

File: clang-tools-extra/clang-tidy/bugprone/AssertSideEffectCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/bugprone/AssertSideEffectCheck.cpp:44]: Merge this FuncDecl with MethodDecl by using dyn_cast_or_null

**Comment by PiotrZSL** [comment]
One more question, there is option for this check:
"CheckFunctionCalls
Whether to treat non-const member and non-member functions as they produce side effects. Disabled by default because it can increase the number of false positive warnings."

Won't this overlap this change ?

File: clang-tools-extra/clang-tidy/bugprone/AssertSideEffectCheck.cpp
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/bugprone/AssertSideEffectCheck.cpp:46]: What if called operator is ->, for example in assert we call -> operator on std::shared_ptr just to get member, and then we call const method, this could cause false-positive.

**Review by PiotrZSL — CHANGES_REQUESTED** [review]
I'm not sure now if this wont cause too much false-positives, after all this is why CheckFunctionCalls option were added. Sometimes object can have 2 same operators, one const and one not, in such case depend on "this" type, non-const could be used, this would be false-positive.

File: clang-tools-extra/clang-tidy/bugprone/AssertSideEffectCheck.cpp
Comment (schenker) [reply clang-tools-extra/clang-tidy/bugprone/AssertSideEffectCheck.cpp]: this makes we wonder if we need the comparisons agains the different operator kinds at all, or if we should treat all operators the same, based on their constness.

File: clang-tools-extra/clang-tidy/bugprone/AssertSideEffectCheck.cpp
Comment (schenker) [reply clang-tools-extra/clang-tidy/bugprone/AssertSideEffectCheck.cpp:46]: I think if the -> operator is implemented correctly, it has to be marked as const, even if it returns something that is not const.

**Comment by schenker** [comment]
> I'm not sure now if this wont cause too much false-positives, after all this is why CheckFunctionCalls option were added. Sometimes object can have 2 same operators, one const and one not, in such case depend on "this" type, non-const could be used, this would be false-positive.

You are right, I misunderstood `CheckFunctionCalls` and it's trickier than I thought. Anyways, I think it would make sense to treat method calls and method operator calls the same way.

I will rething my imlementation, thanks for reviewing!

File: clang-tools-extra/clang-tidy/bugprone/AssertSideEffectCheck.cpp
Comment (PiotrZSL) [reply clang-tools-extra/clang-tidy/bugprone/AssertSideEffectCheck.cpp:46]: On std::shared_ptr yes, but not on std::optional

Comment (schenker) [reply clang-tools-extra/clang-tidy/bugprone/AssertSideEffectCheck.cpp:46]: I see

File: clang-tools-extra/clang-tidy/bugprone/AssertSideEffectCheck.cpp
Comment (schenker) [reply clang-tools-extra/clang-tidy/bugprone/AssertSideEffectCheck.cpp]: Reporting all non const operator calls would not be feasible, because e.g. the `->` operator of `std::optional` is not const but is side-effect free.

**Comment by schenker** [comment]
@PiotrZSL I changed some things, please take another look.

All const operator calls are now considered to be side-effect free. This also fixes some false-positives, e.g. 
```
struct t {
  int operator+=(int i) const { return i; }
} t;
assert(t += 1);
```
is no longer reported.

I added `<<` and `>>` to the list of operators with side-effects. 
Only const calls are reported, so 
```
std::stringstream ss;
assert(ss << 1);
```
is reported, but
```
assert(5 << 1);
```
is not.

There is no overlap with `CheckFunctionCalls`, `CXXOperatorCallExpr`s are not affected by the flag.

File: clang-tools-extra/docs/ReleaseNotes.rst
Comment (PiotrZSL) [inline clang-tools-extra/docs/ReleaseNotes.rst:213]:
```suggestion
  non-const ``<<`` and ``>>`` operators in assertions and fixed some false-positives
```

**Review by PiotrZSL — APPROVED** [review]
LGTM, rebase code before merging

**Review by PiotrZSL — APPROVED** [review]
Code is fine and can be merged, unless you want it merged as "schenker@users.noreply.github.com" you may need to disable email privacy options or try to squash all commits in one.

**Comment by schenker** [comment]
Thanks for pointing out, I squashed and changed the settings.

**Pull request merged**
